Sources/StreamChat/Repositories/AuthenticationRepository.swift

```diff
@@ -135,8 +149,23 @@ class AuthenticationRepository {
     ///   - userInfo:       The user information that will be created OR updated if it exists.
     ///   - tokenProvider:  The block to be used to get a token.
     func connectUser(userInfo: UserInfo?, tokenProvider: @escaping TokenProvider, completion: @escaping (Error?) -> Void) {
-        self.tokenProvider = tokenProvider
-        scheduleTokenFetch(isRetry: false, userInfo: userInfo, tokenProvider: tokenProvider, completion: completion)
+        log.assert(delegate != nil, "Delegate should not be nil at this point")
+
+        let handleTokenFetch = { [weak self] in
+            self?.tokenProvider = tokenProvider
+            self?.scheduleTokenFetch(isRetry: false, userInfo: userInfo, tokenProvider: tokenProvider, completion: completion)
+        }
+
+        let state = EnvironmentState(currentUserId: currentUserId, newUserId: userInfo?.id)
+
+        switch state {
+        case .firstConnection, .newToken:
+            handleTokenFetch()
+        case .newUser:
+            delegate?.logoutUser {
```

Comment:
At this point when we execute `logoutUser` the tokenProvider will be set to the previous value and not the newly provided one. Can that cause an issue?

Reply:
Which previous value? I'm not sure I got it 🤔 The tokenProvider will be the one from the argument, which is the most updated one

Reply:
@ipavlidakis since we will be logging out, during that process we still want to keep the old one. Even if it should not be used, tbh 🙃